plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -648,6 +650,20 @@ evalBuiltinApp fun term runtime = case runtime of
     _ -> pure $ VBuiltin fun term runtime
 {-# INLINE evalBuiltinApp #-}
 
+data ArgsR uni fun ann
+    = NilArgsR
+    | ConsArgsR !(CekValue uni fun ann) !(ArgsR uni fun ann)
+    deriving stock (Show)
+
+toListArgsR :: ArgsR uni fun ann -> [CekValue uni fun ann]
```

Comment:
This isn't on the critical path, so I'd probably just have `ArgsR` be a normal strict list that we remember is in reverse order and stick a `reverse` in the places we materialize it back into a normal-order list. Unsure though.